---
title: "Microarray Data Il1b on stroma cells - GSEA analysis on DGE results"
author: "Hilmar Berger"
output: 
 html_document:
    toc: true
    toc_depth: 3
    toc_float: false
    number_sections: true
    code_folding: hide
    
pdf_document:
    fig_caption: true
date: "`r format(Sys.time(), '%d %B, %Y %H:%M:%S')`"
---

```{r, warning=FALSE, message=FALSE}
rm(list=ls())

library(reshape2)
library(pheatmap)
library(ggplot2)
library(knitr)
library(scales)
library(fgsea)
library(data.table)
library(BiocParallel)
library(readxl)
library(writexl)
#library(org.Mm.eg.db)
library(homologene)

result_folder = "../Results"
if (!file.exists(result_folder)) dir.create(result_folder, recursive=T)

set.seed(12345)
```

# Overview

We here run GSEA on DGE results from microarray of Il1b treated stroma (Giulia Beccaceci). Data was initially analyzed by Hans Mollenkopf (MPIIB). 
For analysis, a preranked analysis using the R package fGSEA is used that should give similar results to preranked analysis in standard GSEA. 

** The naming of the comparisons is target_vs_control. **

# MSigDB gene sets

Since MSigDB gene sets use human gene symbols to map genes to pathways we will translate mouse symbols to homologous human symbols using HomologeneDB from NCBI (build 68). 


<!-- ```{r} -->
<!-- input_files = list.files(path=".", pattern = "DGE_RE ") -->
<!-- names(input_files) = gsub(" vs\\. ", "_vs_", gsub("DGE_RE \\(","", gsub("\\)\\.txt","",input_files))) -->

<!-- all_results = list() -->
<!-- for (n in names(input_files)) { -->
<!--   tmp = read.table(input_files[n], sep="\t", header=T, stringsAsFactors = F) -->
<!--   tmp2 <- strsplit(n, "_") |> unlist() -->
<!--   n_new <- paste(rev(tmp2),collapse="_") -->
<!--   # the files provided by Hans Mollenkopf for some reasons contain the human gene symbols for otherwise mouse data -->
<!--   # We replace these by the correct mouse symbols  -->
<!--   entrez = unique(as.character(tmp$EntrezGeneID)) -->
<!--   entrez = entrez[!is.na(entrez)] -->
<!--   mouse_symbols = select(org.Mm.eg.db, keys=entrez, keytype = "ENTREZID", columns = "SYMBOL") -->
<!--   rownames(mouse_symbols)  = mouse_symbols$ENTREZID -->
<!--   tmp$GeneSymbol = mouse_symbols[as.character(tmp$EntrezGeneID), "SYMBOL"] -->
<!--   all_results[[n_new]] = tmp -->
<!-- } -->


<!-- ``` -->


```{r}
load("../Results/DGE_results.Rdata")
```


```{r}
gsea_gmt_folder = "../Data/External/GeneSets/MSigDB_v7.5"
gsea_gmt_files = c("h.all.v7.5.symbols.gmt","c2.all.v7.5.symbols.gmt","c3.all.v7.5.symbols.gmt","c5.go.bp.v7.5.symbols.gmt","c6.all.v7.5.symbols.gmt","c7.all.v7.5.symbols.gmt","c8.all.v7.5.symbols.gmt")
names(gsea_gmt_files) = c("HALLMARK","PATHWAYS","MOTIF","GO_BP","ONCOGENIC","IMMUNOLOGIC","CELLTYPES")
gset_collections_human = list()
for (n in names(gsea_gmt_files)) {
  gset_collections_human[[n]] = gmtPathways(file.path(gsea_gmt_folder, gsea_gmt_files[[n]]))
}
```

We use gene sets from MSigDB.

```{r}
gsea_gmt_files
```

# Other gene sets 


```{r}
custom_gene_sets_human <- gmtPathways("../Data/External/GeneSets/human_custom_geneset_2025-06-11.gmt" )
custom_gene_sets_mouse <- gmtPathways("../Data/External/GeneSets/mouse_custom_geneset_2025-06-11.gmt" )

names(custom_gene_sets_human)
names(custom_gene_sets_mouse)
```
  - Healthy stromal subset signatures from colon (Mouse colon reference v3)
  - Healthy stromal subset signatures from human stomach (Human stomach reference v1)

```{r, message=FALSE}


# human stomach ref v1
tmp <- as.data.frame(read_excel("../Data/External/GeneSets/scSignatures/Human_stomach_v1/All_cluster_markers_stroma_only_healthy_no_mix.xlsx", sheet=1))
tmp <- subset(tmp, avg_log2FC >= 2.5 & p_val_adj < 0.05)

for (cc in unique(tmp$cluster)) {
  custom_gene_sets_human[[paste0("Healthy_human_stomach_v1_",cc)]] = subset(tmp, cluster == cc)$gene |> unique()
}

# human stomach ref v1, all conditions
tmp <- as.data.frame(read_excel("../Data/External/GeneSets/scSignatures/Human_stomach_v1/All_cluster_markers_stroma_cells.xlsx", sheet=1))
tmp <- subset(tmp, avg_log2FC >= 2.5 & p_val_adj < 0.05 & !cluster %in% c("Epi","n.a.","Mix/Dying","T/NK/B"))

for (cc in unique(tmp$cluster)) {
  custom_gene_sets_human[[paste0("Healthy_human_stomach_v1_all_conditions_",cc)]] = subset(tmp, cluster == cc)$gene |> unique()
}

gset_collections_human[["CUSTOM_HUMAN"]] = custom_gene_sets_human

####################################################################



# mouse colon MCRv3, detailed telocyte
tmp <- as.data.frame(read_excel("../Data/External/GeneSets/scSignatures/MCRv3/All_cluster_markers_subpopulations_all_cells_telocyte_Ctrl_only.xlsx", sheet=1))
tmp <- subset(tmp, avg_log2FC >= 2.5 & p_val_adj < 0.05)

for (cc in unique(tmp$cluster)) {
  custom_gene_sets_mouse[[paste0("Mouse_colon_stroma_MCRv3_healthy_only_",cc)]] = subset(tmp, cluster == cc)$gene |> unique()
}


# mouse colon MCRv3, detailed FBv2
tmp <- as.data.frame(read_excel("../Data/External/GeneSets/scSignatures/MCRv3/All_cluster_markers_subpopulations_all_cells_FBv2_Ctrl_DSS_Inf.xlsx", sheet=1))
tmp <- subset(tmp, avg_log2FC >= 2.5 & p_val_adj < 0.05)

for (cc in unique(tmp$cluster)) {
  custom_gene_sets_mouse[[paste0("Mouse_colon_stroma_MCRv3_FBv2_Ctrl_DSS_",cc)]] = subset(tmp, cluster == cc)$gene |> unique()
}

gset_collections_mouse = list(CUSTOM_MOUSE=custom_gene_sets_mouse)

```

## fGSEA analysis on MSigDB

### Human and Mouse gene sets

For analyzing the human MsigDB gene sets we restrict the DGE data to those probesets that have a homologous gene in mouse and human. For those probesets we select the one with the highest t-score and rank the resulting list which is subsequently used for fGSEA analysis. 

For mouse gene sets (i.e. those derived from mouse experiments) we use the full set of genes in the DGE results after collapsing t-scores by gene and ranking. 

```{r}
data_generation_date = "2026-06-17"
output_folder = result_folder
GSEA_datafile = paste(output_folder,"fGSEA_Results.Rdata",sep="/")
if (!file.exists(output_folder)) dir.create(output_folder, recursive=T)

rerun_GSEA = TRUE

if(rerun_GSEA) {
  all_gsea_results = list()
  n_perm = 5000
  n_proc = 6
  bpparam = MulticoreParam(worker=n_proc, manager.hostname="127.0.0.1")
  for (n in names(all_results)) {
    print(paste("Running FGSEA for comparison",n))
    gsea_result_tables = list()
    tmp = subset(all_results[[n]], !is.na(GeneSymbol)) # those are mostly control probes
    
    m2h <- mouse2human(tmp$GeneSymbol, db = homologeneData2)

    
    tmp2 = merge(tmp, m2h, by.x="GeneSymbol", by.y="mouseGene")
    tmp2 = subset(tmp2, !is.na(humanGene))
    ranks_human = unlist(tapply(tmp2$t, tmp2$humanGene, function(x) x[which(abs(x)==max(abs(x)))] ))

    ranks_mouse = unlist(tapply(tmp$t, tmp$GeneSymbol, function(x) x[which(abs(x)==max(abs(x)))] ))

      
    for (gs in names(gset_collections_human)) {
      print(gs)
      fgseaRes <- fgsea(gset_collections_human[[gs]], ranks_human, minSize=15, maxSize=2000, nperm=n_perm, BPPARAM=bpparam)
      fgseaRes$GeneSetCollection = gs
      gsea_result_tables[[gs]] = fgseaRes
    }
    
    results_human = do.call(rbind, gsea_result_tables)
    
    gsea_result_tables_mouse = list()
    
    for (gs in names(gset_collections_mouse)) {
      print(gs)
      fgseaRes <- fgsea(gset_collections_mouse[[gs]], ranks_mouse, minSize=15, maxSize=2000, nperm=n_perm, BPPARAM=bpparam)
      fgseaRes$GeneSetCollection = gs
      gsea_result_tables_mouse[[gs]] = fgseaRes
    }
  
    results_mouse = do.call(rbind, gsea_result_tables_mouse)
    results_all = rbind(results_human, results_mouse)
    results_all$FDR_global = p.adjust(results_all$pval, "BH")
    
    all_gsea_results[[n]] = results_all
  }
  save(all_gsea_results, file=GSEA_datafile)
  save(gset_collections_human, gset_collections_mouse, file=file.path(output_folder, "Geneset_collections.Rdata"))
} else {
  load(GSEA_datafile)
}
```

Result files are written to folder `r result_folder`

We export all gene sets with less than 20% FDR (false discovery rate) after global adjustment of p-values per sample. 

```{r}
FDR_threshold = 0.5
final_gsea_tabs = list()
if(rerun_GSEA) {
  output_file_prefix = paste(result_folder,"fGSEA_Results_", sep="/")
  for (n in names(all_gsea_results)) {
    filename = paste(output_file_prefix, n, ".txt", sep="" )
    tmp = all_gsea_results[[n]]
    tmp$direction= ifelse(tmp$NES > 0, "Up","Down")
    tmp = tmp[order(tmp$NES, decreasing=T),]
    tmp$leadingEdge = unlist(lapply(tmp$leadingEdge, function(x) paste(x, collapse=",")))
    tmp = subset(tmp, FDR_global < FDR_threshold)
    col_order = c("pathway","GeneSetCollection","direction","pval","padj","ES","NES","nMoreExtreme","size","leadingEdge", "FDR_global")
    write.table(tmp[,col_order, with=F], file= filename, row.names=F , sep="\t", dec=".", quote=F, na="")
    final_gsea_tabs[[n]] = tmp[, col_order, with=F]
  }
  
  DGE_Excel_filename = paste(result_folder,"fGSEA_Results.xlsx",sep="/")
  write_xlsx(final_gsea_tabs, path=DGE_Excel_filename)

}
```



```{r, fig.width=12, fig.height=8}
gsea_summary = list()
for (n in names(all_gsea_results)) {
  tmp = subset(all_gsea_results[[n]], padj < 0.2)
  r = as.data.frame(table(tmp$GeneSetCollection, ifelse(tmp$ES<0, "Down","Up")))
  colnames(r) = c("Collection","Direction","Freq")
  r$Comparison = n
  gsea_summary[[n]] = r
}

gsea_summary_df = do.call(rbind, gsea_summary)

ggplot(gsea_summary_df, aes(x=Comparison, y = Freq, fill=Collection)) + geom_bar(stat="identity") + theme_grey(base_size = 14) + theme(axis.text.x = element_text(angle = 60, hjust = 1)) + ggtitle("Number of significant (FDR<20%) gene sets per comparison")

ggplot(gsea_summary_df, aes(x=Comparison, y = Freq, fill=Collection)) + geom_bar(stat="identity") + theme_grey(base_size = 14) + theme(axis.text.x = element_text(angle = 60, hjust = 1)) + ggtitle("Number of significant (FDR<20%) gene sets per comparison") + facet_wrap(~Direction)
```


```{r}
sessionInfo()
```


